// Copyright 2015, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

syntax = "proto3";


service DisServer {
  // User login
  rpc usrLogin (usrInfo) returns (loginInfo) {}
  // logout
  rpc usrLogout (usrInfo) returns (logoutInfo) {}
  // get all file names
  rpc getFileList (usrInfo) returns (fileList) {}
  // upload
  rpc uploadFile (stream fileData) returns (fileInfo){}
  //download
  rpc downloadFile (fileInfo) returns (stream fileData){}
  //getNodeState
  rpc getState(usrInfo) returns (nodeState){}
}

service CtrServer {
  // User login
  rpc usrLogin (usrInfo) returns (loginInfo) {}
  // logout
  rpc usrLogout (usrInfo) returns (logoutInfo) {}
  // get all file names
  rpc getFileList (usrInfo) returns (fileList) {}
  // upload
  rpc uploadFile (stream fileData) returns (fileInfo){}
  //download
  rpc downloadFile (fileInfo) returns (stream fileData){}
  //node info
  rpc getNodeInfo (usrInfo) returns (nodeInfo){}
}


message usrInfo {//usrLogin Request
  string name = 1;
  string pw = 2;
  string mess = 3;
}

message loginInfo {//usrLogin Reply
  int32 loginFlag = 1;//1 is yes, 0 is no
  string IP = 2;
  int32 accessNum = 3;
  string mess = 4;
}

message logoutInfo {
  int32 logoutFlag = 1;
  string mess = 2;
}

message fileList{
  int32 fileNum = 1;
  repeated string nameList = 2;
  string mess = 3;
}

message fileData{
  string fileName = 1;
  bytes block = 2;
  int32 blockNum = 3;
  string usrName = 4;
  string mess = 5;
}

message fileInfo{
  string fileName = 1;
  int32 blockNum = 2;
  string usrName = 3;
  string mess = 4;
}

message nodeInfo{
  string mess = 1;
}

message nodeState{
  repeated string usrs=1;
  string mess=2;
}
